### PR TITLE
[sw/silicon_creator] Add address translation to the silicon owner stage

### PR DESCRIPTION
**Pull request by engdoreis**
This PR uses the second ibex address translation region to make the silicon owner stage position independent.

(@alphan's update:) @engdoreis, thanks a lot for doing the bulk of the work. As we discussed offline I added some commits to make this ready to merge. A summary of new changes (also described in this [doc](https://docs.google.com/document/d/1T-gQxoR4vRUDYfLe6buJX3UKe49PmOwusfclj41xrEI/edit) that I shared earlier):
* A new virtual memory region `owner_virtual` for the owner stage that starts at `0xa0000000` and is the same size as `rom_ext_virtual`.
* Owner image starts at `owner_virtual` + `ROM_EXT_MAX_SIZE`. Updated the linker script and lma -> vma calculation in `rom_ext.c` accordingly.
* Owner images are now built with proper manifests, i.e. with the correct identifier and `address_translation` value, and are signed.
* ROM_EXT ePMP updates: Removed expectation check since the in-memory state is not initialized yet and minor nit fixes.
* Positive e2e tests for ROM_EXT and the owner stage to have some test coverage for the current behavior.

Created the following issues:
* #13507
* #13511
* #13513

For future reference, various cases can be tested manually using the following commands:
```
# To find the signed binaries for ROM_EXT or the owner stage (substitute `virtual_addr` with `slot_a` or `slot_b` if needed)
bazel build --define DISABLE_VERILATOR_BUILD=true //sw/device/silicon_creator/rom_ext:virtual_addr 2>&1 | grep signed
bazel build --define DISABLE_VERILATOR_BUILD=true //sw/device/silicon_owner/bare_metal:bare_metal_example_virtual_addr 2>&1 | grep signed

# Boot ROM_EXT (path to image file must be absolute):
bazel run //sw/host/opentitantool -- --exec "bootstrap [ROM_EXT_IMAGE]" console

# Boot owner stage:
bazel run //sw/host/opentitantool -- --exec "bootstrap [ROM_EXT_IMAGE]@0 [OWNER_IMAGE]@0x10000" console
```

### CODE REVIEW

File: sw/device/silicon_creator/rom_ext/rom_ext_epmp.h [hunk 1]
```diff
@@ -44,11 +44,25 @@ extern "C" {
...
  */
 void rom_ext_epmp_unlock_owner_stage_rx(epmp_state_t *state,
                                         epmp_region_t image);
 
+/**
+ * Unlocks the provided first silicon owner region with read-only permissions.
+ *
+ * The provided ePMP state is also updated to reflect the changes made to the
+ * hardware configuration.
+ * The image size must be power of 2 as this function uses NAPOT
```
Comment (luismarques) [inline sw/device/silicon_creator/rom_ext/rom_ext_epmp.h:57]:
```suggestion
 * The image size must be a power of 2 as this function uses NAPOT
```

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 2]
```diff
@@ -85,6 +86,25 @@ static rom_error_t rom_ext_verify(const manifest_t *manifest) {
                               &flash_exec);
 }
 
+/* These symbols are defined in
+ * `opentitan/sw/device/silicon_creator/rom_ext/rom_ext.ld`, and describes the
```
Comment (luismarques) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:90]:
```suggestion
 * `opentitan/sw/device/silicon_creator/rom_ext/rom_ext.ld`, and describe the
```

File: sw/device/silicon_owner/bare_metal/bare_metal_virtual_addr.ld [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * Linker script for an OpenTitan first Silicon Owner stage.
+ *
+ * Portions of this file are Ibex-specific.
+ *
+ * The first Silicon Owner stage kept in flash, and can be loaded into either
```
Comment (luismarques) [inline sw/device/silicon_owner/bare_metal/bare_metal_virtual_addr.ld:10]:
```suggestion
 * The first Silicon Owner stage is kept in flash, and can be loaded into either
```

File: sw/device/silicon_owner/bare_metal/bare_metal_virtual_addr.ld [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * Slot A (lower half o the flash), or Slot B (upper half of flash), this
```
Comment (luismarques) [inline sw/device/silicon_owner/bare_metal/bare_metal_virtual_addr.ld:11]:
```suggestion
 * Slot A (lower half o the flash), or Slot B (upper half of flash). This
```

File: sw/device/silicon_owner/bare_metal/bare_metal_virtual_addr.ld [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+/* Reserving space at the top of the RAM for the stack. */
+_stack_size = 0x2000;
+_stack_end = ORIGIN(ram_main) + LENGTH(ram_main);
+_stack_start = _stack_end - _stack_size;
+
+/**
+ * Symbols to be used in the setup of the address translation for ROM_EXT.
+ */
+_slot_start_address = ORIGIN(silicon_owner_virtual);
```
Comment (alphan) [inline sw/device/silicon_owner/bare_metal/bare_metal_virtual_addr.ld:25]: + 64 KiB

File: sw/device/silicon_creator/rom_ext/rom_ext.c [hunk 6]
```diff
@@ -85,6 +86,25 @@ static rom_error_t rom_ext_verify(const manifest_t *manifest) {
...
+extern char _owner_virtual_size[];
+/**
+ * Compute the virtual address corresponding to the physical address `lma_addr`.
+ *
+ * @param manifest Pointer to the current manifest.
+ * @param lma_addr Load address or physical address.
+ * @return the computed virtual address.
+ */
+static uintptr_t owner_vma_get(const manifest_t *manifest, uintptr_t lma_addr) {
+  return (lma_addr - (uintptr_t)manifest +
```
Comment (moidx) [inline sw/device/silicon_creator/rom_ext/rom_ext.c:103]: nit: Add TODO to harden this calculation as an attacker may be able to leverage a data store or bootstrap mechanism to insert malicious code.

Comment (alphan) [reply sw/device/silicon_creator/rom_ext/rom_ext.c:103]: Created #13513 and added a TODO

**Review by cfrantz — APPROVED** [review]
This LGTM.  Just a few nits about comments and names.

File: util/topgen/templates/toplevel_memory.ld.tpl [hunk 7]
```diff
@@ -37,8 +37,8 @@ def flags(mem):
 
 /**
  * Partial linker script for chip memory configuration.
- * eflash virtual is a fixed address that does not physically exist but is used as the
- * translation base
+ * rom_ext_virtual and owner_virtual are fixed addresses that do not physically
+ * exist but is used as the translation base.
```
[the remark this answers is not in the document]
Comment (engdoreis) [reply util/topgen/templates/toplevel_memory.ld.tpl:41]: fixed

**Review by alphan — APPROVED** [review]
@cfrantz, thanks for the review.
@engdoreis, thanks for addressing @cfrantz's comments. We can merge this once all CI jobs pass.

**Comment by alphan** [comment]
Fixed the lint error, `s/flash_name/test_binary` in `sw/device/silicon_owner/bare_metal/BUILD`

**Comment by engdoreis** [comment]
@alphan, I didn't see that you were updating the PR, could push your last change again?

**Comment by alphan** [comment]
> @alphan, I didn't see that you were updating the PR, could push your last change again?

Done

**Pull request merged**